### PR TITLE
[Forms] Added data-mapper docs

### PR DESCRIPTION
**Pull request by vudaltsov**
#SymfonyConHackday2018

Improved version of https://github.com/symfony/symfony-docs/pull/6900
Thank you @wouterj for a great job and the initial text! Thanx @HeahDude for helping me with this PR.

### CODE REVIEW

**Comment by vudaltsov** [comment]
@stof, @xabbuh , need your review :)

File: form/data_mappers.rst [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+and will fit most cases. However, you can create your implementation that
+could, for example, pass submitted data to immutable objects via their constructor.
+
+The Difference between Data Transformers and Mappers
+----------------------------------------------------
+
+It is important to know the difference between
+:doc:`data transformers </form/data_transformers>` and mappers.
+
+* **Data transformers** change the representation of a value. E.g. from
```
Comment (xabbuh) [inline form/data_mappers.rst]: maybe put the example in parentheses:

> change the representation of a value (e.g. from ``"2016-08-12"`` to a ``DateTime`` instance);

File: form/data_mappers.rst [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+The form type should be allowed to edit a color. But because you've decided to
+make the ``Color`` object immutable, a new color object has to be created each time
+one of the values is changed.
+
+.. tip::
+
+    If you're using a mutable object with constructor arguments, instead of
+    using a data mapper, you should configure the ``empty_data`` with a closure
```
Comment (xabbuh) [inline form/data_mappers.rst]: [...] ``empty_data`` option with [...]

File: form/data_mappers.rst [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    as described in
+    :ref:`How to Configure empty Data for a Form Class <forms-empty-data-closure>`.
+
+The red, green and blue form fields have to be mapped to the constructor
+arguments and the ``Color`` instance has to be mapped to red, green and blue
+form fields. Recognize a familiar pattern? It's time for a data mapper!
```
Comment (xabbuh) [inline form/data_mappers.rst]: Maybe end this sentence with a colon.

File: form/data_mappers.rst [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    use App\Painting\Color;
+    use Symfony\Component\Form\DataMapperInterface;
+    use Symfony\Component\Form\Exception\UnexpectedTypeException;
+    use Symfony\Component\Form\FormInterface;
+
+    final class ColorMapper implements DataMapperInterface
+    {
+        /**
+         * {@inheritdoc}
```
Comment (xabbuh) [inline form/data_mappers.rst]: I think we should drop this.

File: form/data_mappers.rst [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            $forms = iterator_to_array($forms);
+
+            // initialize form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        /**
+         * {@inheritdoc}
```
Comment (xabbuh) [inline form/data_mappers.rst]: same here

File: form/data_mappers.rst [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            ;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver
+                ->setDefaults(array(
```
Comment (xabbuh) [inline form/data_mappers.rst]: Use `setDefault()` here instead to simplify the code?

File: form/data_mappers.rst [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            ;
+        }
+    }
+
+Cool! When using the ``ColorType`` form, the custom ``ColorMapper`` will create
+a new ``Color`` object now.
+
+.. caution::
+
+    When a form has ``inherit_data`` option set to ``true``, it does not use the data mapper and
```
Comment (xabbuh) [inline form/data_mappers.rst]: [...] has the `inherit_data` option [...]

File: form/data_mappers.rst [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    lets its parent map inner values.
+
+.. tip::
+
+    You can also implement ``DataMapperInterface`` in the ``ColorType`` and add
```
Comment (xabbuh) [inline form/data_mappers.rst]: [...] implement the `DataMapperInterface` [...]

Comment (xabbuh) [reply form/data_mappers.rst]: Should we also talk about the possibility to implement the data mapper as an anonymous class?

**Review by xabbuh — APPROVED** [review]
great work 👍 just some small minor comments

File: form/data_mappers.rst (hunk 4 above)
Comment (vudaltsov) [reply form/data_mappers.rst]: Agree. @HeahDude , I searched  throughout the docs and found only 2 `{@inheritdoc}` tags. Although I would definitely put them in my project, here in docs they make examples unnecessarily verbose.

File: form/data_mappers.rst (hunk 3 above)
Comment (vudaltsov) [reply form/data_mappers.rst]: I think that `!` brings some more joy and interaction here :)

File: form/data_mappers.rst (hunk 8 above)
Comment (vudaltsov) [reply form/data_mappers.rst]: I think in this very case when we have two rather complicated methods to implement, an anonymous class will be harder to read.

**Comment by xabbuh** [comment]
Thank you @vudaltsov.

**Pull request closed without merging**